### PR TITLE
Custom Secrets Manager and Artifacts registry(dockerhub, ECR, ACR) integration ( Build Env Vars)

### PR DESCRIPTION
**Pull request by lakshya8066**
Added cloud support for Civo and registry support for DockerHub

### CODE REVIEW

File: controllers/api/v1/deploymentset_types.go [hunk 1]
```diff
@@ -24,7 +24,43 @@ import (
...
+}
+
 type DeploymentSetSpec struct {
+	RegistryProvider          string                     `json:"registryProvider,omitempty"`
+	SecretsProvider           string                     `json:"secretsProvider,omitempty"`
+	EcrCredentials            EcrCredentials             `json:"ecrCredentials,omitempty"`
+	DockerHubCredentials      DockerHubCredentials       `json:"dockerHubCredentials,omitempty"`
+	AcrCredentials            AcrCredentials             `json:"acrCredentials,omitempty"`
+	AwsSecretCredentials      AwsSecretCredentials       `json:"awsSecretCredentials,omitempty"`
+	AzureSecretCredentials    AzureSecretCredentials     `json:"azureSecretCredentials,omitempty"`
```
Comment (manav014) [inline controllers/api/v1/deploymentset_types.go]: can be replaced with AzureVaultCredentials

File: agent/constants/constants.go [hunk 2]
```diff
@@ -27,7 +27,43 @@ type IngressYamlManifestType struct {
...
+}
+
 type ParamsConfig struct {
+	SecretsProvider           string
+	EcrCredentials            string
+	DockerHubCredentials      string
+	AcrCredentials            string
+	AwsSecretCredentials      string
+	AzureSecretCredentials    string
+	RegistryProvider          string
```
Comment (manav014) [inline agent/constants/constants.go]: In stead of registry provider is it good to use ArtifactsRegistryProvider?

File: agent/constants/constants.go [hunk 3]
```diff
@@ -27,7 +27,43 @@ type IngressYamlManifestType struct {
...
+type DockerHubCredentials struct {
+	Username   string `json:"username,omitempty"`
+	SecretName string `json:"secretName,omitempty"`
+}
+
+type AcrCredentials struct {
+	AzureManagementScopeToken string `json:"azureManagementScopeToken,omitempty"`
+	AzureAcrRegistryName      string `json:"azureAcrRegistryName,omitempty"`
+	AzureSubscriptionId       string `json:"azureSubscriptionId,omitempty"`
+	AzureResourceGroupName    string `json:"azureResourceGroupName,omitempty"`
```
Comment (manav014) [inline agent/constants/constants.go]: The names for these variables can be changed.

Comment (manav014) [reply agent/constants/constants.go]: No need of Azure prefix when the name is ACR credentials.

File: agent/services/cleanup-kaniko-job-resources.go [hunk 4]
```diff
@@ -18,12 +18,12 @@ func CleanupKanikoJobResources(jobConfig CreateJobConfig) error {
...
-		if err != nil {
-			return err
-		}
-	}
+	// if len(jobConfig.CloudProviderSecretName) != 0 {
+	// 	err := deleteSecret(clientset, jobConfig.CloudProviderSecretName, "humalect")
+	// 	if err != nil {
+	// 		return err
+	// 	}
+	// }
```
Comment (manav014) [inline agent/services/cleanup-kaniko-job-resources.go:27]: If this secret is being used in Application controller then please make sure that it is deleted there.

File: agent/services/create-kaniko-job.go [hunk 5]
```diff
@@ -73,15 +80,130 @@ func CreateKanikoJob(params constants.ParamsConfig) (CreateJobConfig, error) {
...
 func createCloudProviderCredSecrets(clientset *kubernetes.Clientset, params constants.ParamsConfig) (string, error) {
-	if params.CloudProvider == constants.CloudIdAzure {
-		azureCreds, err := getOrgAzureCredsForAcr(params.AzureManagementScopeToken, params.AzureAcrRegistryName,
-			params.AzureSubscriptionId,
-			params.AzureResourceGroupName)
+
+	//----------------------------------ECR REGISTRY---------------------------------------------
+	if params.RegistryProvider == constants.RegistryIdAWS {
+		log.Printf("Ecr Credentials: %v", params.EcrCredentials)
+
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: Never ever log credentials.

File: agent/services/create-kaniko-job.go [hunk 6]
```diff
@@ -73,15 +80,130 @@ func CreateKanikoJob(params constants.ParamsConfig) (CreateJobConfig, error) {
 }
 
 func createCloudProviderCredSecrets(clientset *kubernetes.Clientset, params constants.ParamsConfig) (string, error) {
-	if params.CloudProvider == constants.CloudIdAzure {
-		azureCreds, err := getOrgAzureCredsForAcr(params.AzureManagementScopeToken, params.AzureAcrRegistryName,
-			params.AzureSubscriptionId,
-			params.AzureResourceGroupName)
+
+	//----------------------------------ECR REGISTRY---------------------------------------------
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: No comment needed here.

**Review by manav014 — CHANGES_REQUESTED** [review]

File: agent/services/create-kaniko-job.go
Comment (manav014) [inline agent/services/create-kaniko-job.go:1]: Can We Make the changes in this file more modular maybe we can move a few functions to specific services/utils.

**Review by manav014 — CHANGES_REQUESTED** [review]

File: agent/tasks/deploy.go [hunk 7]
```diff
@@ -37,6 +39,9 @@ func Deploy(config *constants.ParamsConfig) error {
 	// 	fmt.Println(err)
 	// 	return err
 	// }
+	log.Println("Creating Docker Registry Secret AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")
+	fmt.Println("Creating Docker  dfas gRegistry Se sagcret AAAAAsdgadbAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")
+
```
Comment (manav014) [inline agent/tasks/deploy.go]: No need of logs here.

File: controllers/internal/controller/deploymentset_controller.go [hunk 8]
```diff
@@ -153,9 +190,16 @@ func (r *DeploymentSetReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 					Containers: []corev1.Container{
 						{
 							Name:            fmt.Sprintf("%s-ds-%s", deploymentSet.Spec.ManagedBy, deploymentSet.Spec.CommitId),
-							Image:           "public.ecr.aws/humalect/core-agent:" + agentImageTag,
+							Image:           "public.ecr.aws/humalect/newagent:lakshya",
```
Comment (lakshya8066) [inline controllers/internal/controller/deploymentset_controller.go]: *****Have to revert this****

**Review by manav014 — CHANGES_REQUESTED** [review]

**Comment by manav014** [comment]
Please delete the secrets required by K8s deployment to pull the docker image.

File: agent/tasks/deploy.go [hunk 9]
```diff
@@ -89,7 +92,25 @@ func Deploy(config *constants.ParamsConfig) error {
...
+		services.SendWebhook(config.WebhookEndpoint, config.WebhookData, false, constants.CreatedApplicationCrd)
+		return err
+	}
+
+	var azureVaultCredentials constants.AzureVaultCredentials
+	err = json.Unmarshal([]byte(config.AzureVaultCredentials), &azureVaultCredentials)
+
+	if err != nil {
+		fmt.Println(err)
+		return err
```
Comment (manav014) [inline agent/tasks/deploy.go]: send webhook here as well.

File: agent/utils/get-cli-arguments.go [hunk 10]
```diff
@@ -8,11 +8,18 @@ import (
...
+	flag.StringVar(&config.EcrCredentials, "ecrCredentials", "", "This is an optional parameter and it would only be passed when the artifactsRegistryProvider is aws and it represents the credentials for the AWS ECR registry.")
+	flag.StringVar(&config.DockerHubCredentials, "dockerHubCredentials", "", "This is an optional parameter and it would only be passed when the artifactsRegistryProvider is dockerhub and it represents the credentials for the dockerhub registry.")
+	flag.StringVar(&config.AcrCredentials, "acrCredentials", "", "This is an optional parameter and it would only be passed when the artifactsRegistryProvider is azure and it represents the credentials for the Azure ACR registry.")
+	flag.StringVar(&config.AwsSecretCredentials, "awsSecretCredentials", "", "This is an optional parameter and it would only be passed when the artifactsRegistryProvider is aws and it represents the credentials for the AWS ECR registry.")
+	flag.StringVar(&config.AzureVaultCredentials, "azureVaultCredentials", "", "This is an optional parameter and it would only be passed when the artifactsRegistryProvider is azure and it represents the credentials for the Azure ACR registry.")
```
Comment (manav014) [inline agent/utils/get-cli-arguments.go]: Change the messages here

File: agent/utils/get-cli-arguments.go [hunk 11]
```diff
@@ -8,11 +8,18 @@ import (
...
 	flag.StringVar(&config.AzureManagementScopeToken, "azureManagementScopeToken", "", "This is the optional parameter which shows the token of Azure for https://management.azure.com/.default scope and this would only be passed when the cloudProvider is azure.")
 	flag.StringVar(&config.SourceCodeRepositoryName, "sourceCodeRepositoryName", "", "The repository name of your github, gitlab or bitbucket repository")
 	flag.StringVar(&config.SourceCodeProvider, "sourceCodeProvider", "", "This is a required parameter that represents the name of the version control system portal and it could be either github  or gitlab or bitbucket")
 	flag.StringVar(&config.SourceCodeToken, "sourceCodeToken", "", "This is a required parameter that represents the token to pull the source code from the version control system portal namely github, gitlab and bitbucket.")
+	flag.StringVar(&config.ArtifactsRegistryProvider, "artifactsRegistryProvider", "", "This is a required parameter that represents the name of the docker registry provider and it could be either dockerhub or azure or aws.")
```
Comment (manav014) [inline agent/utils/get-cli-arguments.go:21]: Change message here no required as well.

File: agent/utils/get-cli-arguments.go [hunk 12]
```diff
@@ -8,11 +8,18 @@ import (
 
 func ParseCLIArguments() *constants.ParamsConfig {
 	config := &constants.ParamsConfig{}
+	flag.StringVar(&config.SecretsProvider, "secretsProvider", "", "This is a required parameter which can have 2 values either aws or azure and it represents the main secrets provider to be used.")
```
Comment (manav014) [inline agent/utils/get-cli-arguments.go]: Optional

File: agent/services/create-kaniko-job.go [hunk 13]
```diff
@@ -73,18 +80,127 @@ func CreateKanikoJob(params constants.ParamsConfig) (CreateJobConfig, error) {
...
-			params.AzureSubscriptionId,
-			params.AzureResourceGroupName)
+	if params.ArtifactsRegistryProvider == constants.RegistryIdAWS {
+		var creds map[string]interface{}
+		err := json.Unmarshal([]byte(params.EcrCredentials), &creds)
+
+		if err != nil {
+			log.Fatalf("Failed to parse ECR credentials got error : %v", err)
+			return "", err
+		}
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: Make a separate function to unmarshall stuff.

File: agent/services/create-kaniko-job.go [hunk 14]
```diff
@@ -73,18 +80,127 @@ func CreateKanikoJob(params constants.ParamsConfig) (CreateJobConfig, error) {
 }
 
 func createCloudProviderCredSecrets(clientset *kubernetes.Clientset, params constants.ParamsConfig) (string, error) {
-	if params.CloudProvider == constants.CloudIdAzure {
-		azureCreds, err := getOrgAzureCredsForAcr(params.AzureManagementScopeToken, params.AzureAcrRegistryName,
-			params.AzureSubscriptionId,
-			params.AzureResourceGroupName)
+	if params.ArtifactsRegistryProvider == constants.RegistryIdAWS {
+		var creds map[string]interface{}
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: in stead of map[strin]interface{} use specific type defined in constants.

File: agent/services/create-kaniko-job.go [hunk 15]
```diff
@@ -73,18 +80,127 @@ func CreateKanikoJob(params constants.ParamsConfig) (CreateJobConfig, error) {
...
+		err := json.Unmarshal([]byte(params.EcrCredentials), &creds)
+
+		if err != nil {
+			log.Fatalf("Failed to parse ECR credentials got error : %v", err)
+			return "", err
+		}
+
+		ecrToken, err := getEcrLoginToken(creds["accessKey"].(string), creds["secretKey"].(string))
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: Change here as well once you will unmarshall to a type.

File: agent/services/create-kaniko-job.go [hunk 16]
```diff
@@ -397,3 +525,181 @@ func createKanikoConfigResources(clientset *kubernetes.Clientset, params constan
...
+		return "", fmt.Errorf("value not found in response JSON")
+	}
+
+	return secretValue, nil
+}
+
+func getEcrLoginToken(accessKey string, secretKey string) (string, error) {
+
+	sess, err := session.NewSession(&aws.Config{
+		Region:      aws.String("ap-south-1"),
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: Don't hardcode region here take it from user.

File: agent/services/create-kaniko-job.go [hunk 17]
```diff
@@ -73,18 +80,127 @@ func CreateKanikoJob(params constants.ParamsConfig) (CreateJobConfig, error) {
...
+		}
+
+		dockerRegistrySecret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: fmt.Sprintf("%s-ksec-ecr-%s-%s",
+					params.ManagedBy[:int(math.Min(float64(len(params.ManagedBy)), float64(10)))],
+					params.CommitId[:int(math.Min(float64(len(params.CommitId)), float64(5)))],
+					params.DeploymentId[:int(math.Min(float64(len(params.DeploymentId)), float64(7)))]),
+				Namespace: "humalect",
+				Labels: map[string]string{
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: change label here to match it with name

File: agent/services/create-kaniko-job.go [hunk 18]
```diff
@@ -73,18 +80,127 @@ func CreateKanikoJob(params constants.ParamsConfig) (CreateJobConfig, error) {
...
+					"ManagedBy":    params.ManagedBy,
+					"CommitId":     params.CommitId,
+				},
+			},
+			Type: corev1.SecretTypeDockerConfigJson,
+			StringData: map[string]string{
+				".dockerconfigjson": fmt.Sprintf(`{  
+					"auths": {  
+						"%s": {  
+							"username":"AWS",
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: Check if in stead of hardcoding this, can this be dynamic

File: agent/services/create-kaniko-job.go [hunk 19]
```diff
@@ -397,3 +525,181 @@ func createKanikoConfigResources(clientset *kubernetes.Clientset, params constan
...
+
+	password := strings.Split(string(decodedToken), ":")[1]
+	fmt.Println("ECR login password:", password)
+
+	return password, nil
+}
+
+func getDockerHubSecretKey(params constants.ParamsConfig) (string, error) {
+
+	var dockerHubCreds map[string]interface{}
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: unmarshall to a constant type.

File: agent/services/create-kaniko-job.go [hunk 20]
```diff
@@ -397,3 +525,181 @@ func createKanikoConfigResources(clientset *kubernetes.Clientset, params constan
...
+	err := json.Unmarshal([]byte(params.DockerHubCredentials), &dockerHubCreds)
+
+	if err != nil {
+		log.Fatalf("Failed to parse ECR credentials got error : %v", err)
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: message check

File: agent/services/create-kaniko-job.go [hunk 21]
```diff
@@ -397,3 +525,181 @@ func createKanikoConfigResources(clientset *kubernetes.Clientset, params constan
 	}
 	return CreateJobConfig{CloudProviderSecretName: cloudProviderSecretName, DockerFileConfigName: dockerFileConfigName}, nil
 }
+
+func getAwsSecretValue(secretName, accessKey, secretKey, region string) (string, error) {
+	// Create a session object with the access key and secret key
+	sess, err := session.NewSession(&aws.Config{
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: Redundant code.

File: agent/services/create-kaniko-job.go [hunk 22]
```diff
@@ -397,3 +525,181 @@ func createKanikoConfigResources(clientset *kubernetes.Clientset, params constan
...
+
+	var secretData map[string]string
+	err = json.Unmarshal([]byte(secretValue), &secretData)
+	if err != nil {
+		fmt.Println("Error unmarshalling JSON:", err)
+		return "", err
+	}
+
+	// Return the "dockerhub" key's value
+	return secretData["dockerhub"], nil
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: make it a constant.

File: agent/services/create-kaniko-job.go [hunk 23]
```diff
@@ -397,3 +525,181 @@ func createKanikoConfigResources(clientset *kubernetes.Clientset, params constan
...
+	if err != nil {
+		return "", fmt.Errorf("error unmarshalling response JSON: %v", err)
+	}
+
+	secretValue, ok := responseJSON["value"].(string)
+	if !ok {
+		return "", fmt.Errorf("value not found in response JSON")
+	}
+
+	return secretValue, nil
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: Use the same key to fetch secret value here same as that of AWS.
 that is ["dockerhub"]

File: agent/services/create-kaniko-job.go [hunk 24]
```diff
@@ -397,3 +525,181 @@ func createKanikoConfigResources(clientset *kubernetes.Clientset, params constan
...
+		err := json.Unmarshal([]byte(params.AwsSecretCredentials), &awsCreds)
+
+		if err != nil {
+			log.Fatalf("Failed to parse Aws credentials got error : %v", err)
+			return "", err
+		}
+		return getAwsSecretValue(dockerHubCreds["secretName"].(string), awsCreds["accessKey"].(string), awsCreds["secretKey"].(string), awsCreds["region"].(string))
+	} else if params.SecretsProvider == constants.CloudIdAzure {
+
+		var azureCreds map[string]interface{}
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: variable name can be azureVaultCreds.

File: agent/services/create-kaniko-job.go [hunk 25]
```diff
@@ -397,3 +525,181 @@ func createKanikoConfigResources(clientset *kubernetes.Clientset, params constan
...
+		err := json.Unmarshal([]byte(params.AzureVaultCredentials), &azureCreds)
+
+		if err != nil {
+			log.Fatalf("Failed to parse Aws credentials got error : %v", err)
+			return "", err
+		}
+		secretData, err := getAzureSecretString(azureCreds["token"].(string), azureCreds["name"].(string), dockerHubCreds["secretName"].(string))
+		// secretData, err := getAzureSecretString(params.AzureVaultToken, params.AzureVaultName, dockerHubCreds["secretName"].(string))
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: please remove this comment.

File: agent/services/create-kaniko-job.go [hunk 26]
```diff
@@ -73,18 +80,127 @@ func CreateKanikoJob(params constants.ParamsConfig) (CreateJobConfig, error) {
...
+						"%s": {  
+							"username":"AWS",
+							"password": "%s"  
+						}  
+					}  
+				}`, creds["registryUrl"], ecrToken),
+			},
+		}
+
+		createdSecret, err := clientset.CoreV1().Secrets("humalect").Create(context.Background(), dockerRegistrySecret, metav1.CreateOptions{})
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: Creating this in user provided namespace would be better IMO

File: agent/services/create-kaniko-job.go [hunk 27]
```diff
@@ -73,18 +80,127 @@ func CreateKanikoJob(params constants.ParamsConfig) (CreateJobConfig, error) {
...
+			return "", err
+		}
+
+		dockerRegistrySecret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: fmt.Sprintf("%s-ksec-ecr-%s-%s",
+					params.ManagedBy[:int(math.Min(float64(len(params.ManagedBy)), float64(10)))],
+					params.CommitId[:int(math.Min(float64(len(params.CommitId)), float64(5)))],
+					params.DeploymentId[:int(math.Min(float64(len(params.DeploymentId)), float64(7)))]),
+				Namespace: "humalect",
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: Creating this in user provided namespace would be better IMO

File: agent/services/create-kaniko-job.go (hunk 26 above)
Comment (manav014) [reply agent/services/create-kaniko-job.go]: Do this for all the secrets created int his namespace.

File: agent/services/create-kaniko-job.go [hunk 28]
```diff
@@ -73,18 +80,127 @@ func CreateKanikoJob(params constants.ParamsConfig) (CreateJobConfig, error) {
...
+		}
+
+		ecrToken, err := getEcrLoginToken(creds["accessKey"].(string), creds["secretKey"].(string))
+
+		if err != nil {
+			log.Fatalf("Error getting ECR token: %v", err)
+			return "", err
+		}
+
+		dockerRegistrySecret := &corev1.Secret{
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: you are creating this 3 times using same redundant code please optimize that by using secretstring as a variable.

File: agent/services/create-kaniko-job.go [hunk 29]
```diff
@@ -233,12 +344,29 @@ func getKanikoJobObject(
...
+		var dockerHubCreds map[string]interface{}
+		err := json.Unmarshal([]byte(params.DockerHubCredentials), &dockerHubCreds)
+		if err != nil {
+			log.Fatalf("Failed to parse ECR credentials got error : %v", err)
+		}
+		artifactsRepoUrl = fmt.Sprintf("%s/%s:%s", dockerHubCreds["username"], params.ArtifactsRepositoryName, params.CommitId)
+	} else {
+
+		var creds map[string]interface{}
+		err := json.Unmarshal([]byte(params.EcrCredentials), &creds)
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: Please avoid redundancy in case of unmarshal.

File: agent/services/create-kaniko-job.go [hunk 30]
```diff
@@ -397,3 +525,181 @@ func createKanikoConfigResources(clientset *kubernetes.Clientset, params constan
...
+	}
+
+	decodedToken, err := base64.StdEncoding.DecodeString(*result.AuthorizationData[0].AuthorizationToken)
+	if err != nil {
+		fmt.Println("Error decoding ECR authorization token:", err)
+		return "", err
+	}
+
+	password := strings.Split(string(decodedToken), ":")[1]
+	fmt.Println("ECR login password:", password)
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: Remove this.

File: agent/services/create-kaniko-job.go [hunk 31]
```diff
@@ -233,12 +344,29 @@ func getKanikoJobObject(
...
 ) (batchv1.Job, error) {
 	var artifactsRepoUrl string
-	if params.CloudProvider == constants.CloudIdAzure {
+	if params.CloudProvider == constants.CloudIdAzure || params.ArtifactsRegistryProvider == constants.RegistryIdAzure {
 		artifactsRepoUrl = fmt.Sprintf("%s.azurecr.io/%s:%s", params.AzureAcrRegistryName, params.
 			ArtifactsRepositoryName, params.CommitId)
 	} else if params.CloudProvider == constants.CloudIdAWS {
 		artifactsRepoUrl = fmt.Sprintf("%s/%s:%s", params.AwsEcrRegistryUrl, params.
 			ArtifactsRepositoryName, params.CommitId)
+	} else if params.CloudProvider == constants.CloudIdCivo && params.ArtifactsRegistryProvider == constants.RegistryIdDockerhub {
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: Reduce these conditions and make them Artifacts registry specific.

Comment (manav014) [reply agent/services/create-kaniko-job.go]: use them from Acrcredentials and azurecredentials.

File: controllers/api/v1/deploymentset_types.go
Comment (manav014) [inline controllers/api/v1/deploymentset_types.go:1]: Remove unnecessary unused variables.

File: controllers/api/v1/application_types.go
Comment (manav014) [inline controllers/api/v1/application_types.go:1]: Remove unnecessary unused variables.

File: agent/services/create-kaniko-job.go (hunk 29 above)
Comment (manav014) [reply agent/services/create-kaniko-job.go]: Make ArtifactsRepositoryName required in types

File: controllers/internal/controller/deploymentset_controller.go [hunk 32]
```diff
@@ -134,11 +134,48 @@ func (r *DeploymentSetReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 	log.Info("Creating Job")
 	fmt.Println(deploymentSet.Spec.DockerManifest)
 	agentImageTag, exists := os.LookupEnv("AGENT_IMAGE_TAG")
+
 	if !exists {
 		agentImageTag = "latest"
+		fmt.Print(agentImageTag)
```
Comment (manav014) [inline controllers/internal/controller/deploymentset_controller.go:142]: please no

**Review by manav014 — CHANGES_REQUESTED** [review]

File: controllers/internal/controller/deploymentset_controller.go [hunk 33]
```diff
@@ -134,11 +134,48 @@ func (r *DeploymentSetReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+		panic(err)
+	}
+
+	azureVaultCredentials, err := json.Marshal(deploymentSet.Spec.AzureVaultCredentials)
+	if err != nil {
+		deploymentSet.Spec.WebhookData = helpers.UpdateStatusData(deploymentSet.Spec.WebhookData, constants.DeploymentJobCreated, false)
+		sendDeploymentJobCreatedWebhook(*deploymentSet, false)
+		panic(err)
+	}
+
```
Comment (manav014) [inline controllers/internal/controller/deploymentset_controller.go]: Too much of redundant if can we move this to a function.

File: controllers/internal/controller/deploymentset_controller.go (hunk 8 above)
Comment (manav014) [inline controllers/internal/controller/deploymentset_controller.go]: Please remove this.

File: controllers/internal/controller/application_creation.go [hunk 34]
```diff
@@ -44,7 +44,9 @@ func (r *ApplicationReconciler) handleCreation(ctx context.Context, application
 			},
 			Namespace: Namespace,
 		}
-		SecretStringData, err := cloudhelpers.GetCloudSecretMap(application.Spec.AzureVaultToken, application.Spec.SecretManagerName, application.Spec.CloudRegion, application.Spec.AzureVaultName, application.Spec.CloudProvider)
+		SecretStringData, err := cloudhelpers.GetCloudSecretMap(application.Spec.SecretsProvider, application.Spec.AwsSecretCredentials.AccessKey,
+			application.Spec.AwsSecretCredentials.SecretKey, application.Spec.AwsSecretCredentials.Region, application.Spec.AzureVaultCredentials.Token,
+			application.Spec.SecretManagerName, application.Spec.CloudRegion, application.Spec.AzureVaultCredentials.Name, application.Spec.CloudProvider)
```
Comment (manav014) [inline controllers/internal/controller/application_creation.go]: Send complete credentials struct here.

Comment (manav014) [reply controllers/internal/controller/application_creation.go]: Just send application.

File: controllers/internal/controller/helpers/cloud/secrets.go [hunk 35]
```diff
@@ -13,25 +13,44 @@ import (
...
+	} else if cloudId == constants.CloudIdCivo && secretsProvider == constants.CloudIdAWS {
+		// Load custom configuration
+		config, err := config.LoadDefaultConfig(context.TODO(), config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(awsAccessKey, awsSecretKey, "")))
+		if err != nil {
+			log.Fatal(err)
+			return map[string]string{}, err
+		}
+
+		secretString, err = getAwsSecretString(config, secretName, awsRegion)
+
```
Comment (manav014) [inline controllers/internal/controller/helpers/cloud/secrets.go]: if secret provider aws and azure that's it and handle cases.

**Review by manav014 — CHANGES_REQUESTED** [review]

File: agent/services/create-kaniko-job.go (hunk 17 above)
Comment (lakshya8066) [reply agent/services/create-kaniko-job.go]: shouldn't it be the name of the app it is associated with?

File: controllers/internal/controller/deploymentset_controller.go (hunk 8 above)
Comment (lakshya8066) [reply controllers/internal/controller/deploymentset_controller.go]: Kept this for testing purposes, will remove this right before the pr is ready to be merged

File: agent/utils/get-cli-arguments.go (hunk 12 above)
Comment (lakshya8066) [reply agent/utils/get-cli-arguments.go]: I think this should be required even if secrets provider is the same as the cloud provider, for better **isolation**

File: agent/services/create-kaniko-job.go [hunk 36]
```diff
@@ -73,58 +80,71 @@ func CreateKanikoJob(params constants.ParamsConfig) (CreateJobConfig, error) {
...
-		createdSecret, err := clientset.CoreV1().Secrets("humalect").Create(context.Background(), dockerRegistrySecret, metav1.CreateOptions{})
+		secretData = fmt.Sprintf(`{  
+			"auths": {  
+				"%s": {  
+					"username":"AWS",
+					"password": "%s"  
+				}  
+			}  
+		}`, ecrCredentials.RegistryUrl, ecrToken)
+		return createKanikoSecret(secretData, params.ArtifactsRegistryProvider, params, clientset)
```
Comment (manav014) [inline agent/services/create-kaniko-job.go]: When you are calling it outside then what is the need to call it here.

**Review by manav014 — CHANGES_REQUESTED** [review]

File: controllers/internal/controller/helpers/cloud/secrets.go [hunk 37]
```diff
@@ -10,30 +10,35 @@ import (
...
 			return map[string]string{}, err
 		}
-	} else if cloudId == constants.CloudIdAzure {
-		secretString, err = getAzureSecretString(azureVaultToken, vaultName, secretName)
+
+		secretString, err = getAwsSecretString(config, application.Spec.SecretManagerName, application.Spec.AwsSecretCredentials.Region)
+
+	} else if application.Spec.SecretsProvider == constants.CloudIdAzure || (application.Spec.SecretsProvider == "" && application.Spec.CloudProvider == constants.CloudIdAzure) {
+		secretString, err = getAzureSecretString(application.Spec.AzureVaultCredentials.Token, application.Spec.AzureVaultCredentials.Name, application.Spec.SecretManagerName)
 		if err != nil {
```
Comment (manav014) [inline controllers/internal/controller/helpers/cloud/secrets.go]: This is incomplete

**Review by manav014 — CHANGES_REQUESTED** [review]

File: controllers/internal/controller/helpers/cloud/secrets.go (hunk 37 above)
Comment (manav014) [reply controllers/internal/controller/helpers/cloud/secrets.go]: In `services/aws/get-secret-value.go` we are using session while here, we are using config. for secret fetching. Let's keep both of them consistent.

Comment (manav014) [reply controllers/internal/controller/helpers/cloud/secrets.go]: Make it compatible such that if the cluster if AWS and the secretProviderCredentials is not provided then it should work properly as node group has access to secrets

File: agent/services/dockerhub/fetch-secret-key.go [hunk 38]
```diff
@@ -12,10 +12,10 @@ import (
 
 func FetchDockerHubSecretKey(params constants.ParamsConfig) (string, error) {
 	dockerHubCreds := utils.UnmarshalStrings(params.DockerHubCredentials).(constants.DockerHubCredentials)
-	if params.SecretsProvider == constants.CloudIdAWS {
+	if params.SecretsProvider == constants.CloudIdAWS || (params.SecretsProvider == "" && params.CloudProvider == constants.CloudIdAWS) {
```
Comment (lakshya8066) [inline agent/services/dockerhub/fetch-secret-key.go:16]: Good catch!

**Review by vishalsaugat — APPROVED** [review]

**Pull request merged**